File: python/0047.py
class Solution(object):
    def __init__(self):
        self.ans=[]
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        def go(i):
            if i==len(nums):
                self.ans.append(nums[:])
                return
            ma=set()
            for j in range(i,len(nums)):
                if nums[j] not in ma:
                    ma.add(nums[j])
                    nums[i],nums[j]=nums[j],nums[i]
                    go(i+1)
                    nums[i],nums[j]=nums[j],nums[i]
        nums=sorted(nums)
        go(0)
        return self.ans

File: rust/src/bin/0047.rs
struct Solution;


fn go(ans: &mut Vec<Vec<i32>>, ind: usize, a: &mut Vec<i32>) {
    // println!("ind={},a={:?}", ind, a);
    if ind == a.len() {
        ans.push(a.clone());
        return;
    }

    for i in ind..a.len() {
        a.swap(ind, i);
        go(ans, ind + 1, a);
        a.swap(ind, i);
    }
}

impl Solution {
    pub fn permute_unique(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort();
        let mut ans: Vec<Vec<i32>> = vec![];
        go(&mut ans, 0, &mut nums);
        ans.sort();
        if ans.len() == 0 {
            return vec![];
        }
        let mut valid = vec!(ans[0].clone());
        for i in 1..ans.len() {
            if ans[i] != ans[i - 1] {
                valid.push(ans[i].clone());
            }
        }
        valid
    }
}


fn main() {
    let ans = Solution::permute_unique(vec!(1, 1, 2));
    println!("{:?}", ans);
}